Docs: Adds plugin signature verification pages (#25257)
grafana/grafana: docs/sources/developers/plugins/signing.md

+++
title = "Signing"
type = "docs"
+++

# Signing

> Signing is an additional security measure to make sure plugins you authored, installed or using haven't been tampered with. Grafana will verify whether a plugin is signed or unsigned when starting up and loading plugins by inspecting and verifying its digital signature, if any.

Reviewer: ```suggestion
Plugin signing is a security measure to make sure plugins haven't been tampered with. Grafana will verify if a plugin is signed or unsigned when loading it by inspecting and verifying its digital signature.
```

> Note: All Grafana Labs authored backend plugins, including Enterprise plugins, are signed. We're looking into providing a process for allowing  community plugins to be signed in an upcoming version of Grafana.

## Signing a plugin

> Signing a plugin is the process of creating a signed manifest file, _MANIFEST.txt_. The first step is to create the plugin manifest. This is a JSON file that should contain the same plugin id and version found in `plugin.json`. The list of plugin files with their respective checksums (SHA256) should alow be included as `files`. Example plugin manifest:

Reviewer: ```suggestion
Signing a plugin is the process of creating a signed manifest file, _MANIFEST.txt_. The first step is to create the plugin manifest. This is a JSON file that should contain the same plugin id and version found in `plugin.json`. The list of plugin files with their respective checksums (SHA256) should all be included as `files`. Example plugin manifest:
```

> Signing a plugin is the process of creating a signed manifest file, _MANIFEST.txt_. The first step is to create the plugin manifest. This is a JSON file that should contain the same plugin id and version found in `plugin.json`. The list of plugin files with their respective checksums (SHA256) should alow be included as `files`. Example plugin manifest:

Author: You would use grafana toolkit to create the manifest JSON, sign it and create it by calling grafana.com API.

> Signing a plugin is the process of creating a signed manifest file, _MANIFEST.txt_. The first step is to create the plugin manifest. This is a JSON file that should contain the same plugin id and version found in `plugin.json`. The list of plugin files with their respective checksums (SHA256) should alow be included as `files`. Example plugin manifest:

Author: So idea was not to be task, rather explains how plugins are signed on a higher level.

> Signing a plugin is the process of creating a signed manifest file, _MANIFEST.txt_. The first step is to create the plugin manifest. This is a JSON file that should contain the same plugin id and version found in `plugin.json`. The list of plugin files with their respective checksums (SHA256) should alow be included as `files`. Example plugin manifest:

Reviewer 2: Hm. Then I'm not sure who's the intended audience of this page. 

If it's for end-users who are wondering what the `Signed` label stands for, then I don't think we need to explain signing to this extent, and we should put it in the general plugin documentation.

If it's for plugin developers, then we can probably leave it at explaining that it's a mechanism for Grafana-authored plugin, since they can't really do anything with the information?

Who needs to read this document, and why?

> Signing a plugin is the process of creating a signed manifest file, _MANIFEST.txt_. The first step is to create the plugin manifest. This is a JSON file that should contain the same plugin id and version found in `plugin.json`. The list of plugin files with their respective checksums (SHA256) should alow be included as `files`. Example plugin manifest:

Author: It's purpose is to bridge the gap between the understand of and Grafana logging error "plugin is unsigned" and https://grafana.com/docs/grafana/latest/installation/configuration/#allow-loading-unsigned-plugins

> Signing a plugin is the process of creating a signed manifest file, _MANIFEST.txt_. The first step is to create the plugin manifest. This is a JSON file that should contain the same plugin id and version found in `plugin.json`. The list of plugin files with their respective checksums (SHA256) should alow be included as `files`. Example plugin manifest:

Author: What do you suggest, that we move some parts to https://grafana.com/docs/grafana/latest/plugins/ and keep the plugin developer specific information here?

> Signing a plugin is the process of creating a signed manifest file, _MANIFEST.txt_. The first step is to create the plugin manifest. This is a JSON file that should contain the same plugin id and version found in `plugin.json`. The list of plugin files with their respective checksums (SHA256) should alow be included as `files`. Example plugin manifest:

Reviewer 2: I think that might be a good idea. Basically have a section under the general plugin category:

```
Plugins
  Overview
  Install plugins
  Signed plugins
```

that explains the what and why on a higher level. This is useful information for anyone using plugins in Grafana.

For the docs aimed at plugin developers, given that they can't sign plugins themselves, the only information that's interesting to a plugin developer at this time is: "What can/should I do in the meantime?". I don't think we need to add more details than that. 

@oddlittlebird?

> Signing a plugin is the process of creating a signed manifest file, _MANIFEST.txt_. The first step is to create the plugin manifest. This is a JSON file that should contain the same plugin id and version found in `plugin.json`. The list of plugin files with their respective checksums (SHA256) should alow be included as `files`. Example plugin manifest:

Author: Have split up in two pages. Please have a look

```json
{
  "plugin": "grafana-test-plugin",
  "version": "1.0.0",
  "files": {
    "LICENSE": "cfc7749b96f63bd31c3c42b5c471bf756814053e847c10f3eb003417bc523d30",
    "README.md": "5bfefcdce6eafce3388d1fb200f3b10954cfeac6c7a45fd7dec42687e01ac75d",
    "module.js": "3c07596a6a4796c65ef10ba2bc0805e7f3bc9e4e8fc9970a1307b97e29db1c0a",
    "module.js.LICENSE.txt": "fdbc28c10f3d21976b4bc16464ad7c630538c0c3101347b5fd44af9066f7022b",
    "module.js.map": "c3ac1e8aa98d83c54fd13e43b65e1cf4182a924d2eb23a2f1a6fe40b7785a1bb",
    "plugin.json": "cf26a3afb7c10cd9ae40b5296d04172b5dac927d69a51082e6d085b34341ccc3"
  }
}
```

> Next the plugin manifest is signed with a private key and create the _MANIFEST.txt_. The manifest signing process is offloaded to the grafana.com HTTP API and is currently only available for Grafana Labs authored plugins. The HTTP API result is stored as _MANIFEST.txt_ side by side with `plugin.json` and other plugin files. The last step is to archive the plugin as a zip-file.

Author: ```suggestion
Next, the plugin manifest is signed with a private key and the _MANIFEST.txt_ is created. The manifest signing process is offloaded to the grafana.com HTTP API and is currently only available for Grafana Labs authored plugins. The HTTP API result is stored as _MANIFEST.txt_ side by side with `plugin.json` and other plugin files. The last step is to archive the plugin as a zip-file.
```

> Next the plugin manifest is signed with a private key and create the _MANIFEST.txt_. The manifest signing process is offloaded to the grafana.com HTTP API and is currently only available for Grafana Labs authored plugins. The HTTP API result is stored as _MANIFEST.txt_ side by side with `plugin.json` and other plugin files. The last step is to archive the plugin as a zip-file.

Reviewer: "Next the plugin manifest is signed with a private key and create the _MANIFEST.txt_" ?

Having trouble parsing this.

Next, the plugin manifest is signed with a private key and the _MANIFEST.TEXT_ is created. ...?

```txt
// MANIFEST.txt
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

{
  "plugin": "grafana-test-plugin",
  "version": "1.0.0",
  "files": {
    "LICENSE": "cfc7749b96f63bd31c3c42b5c471bf756814053e847c10f3eb003417bc523d30",
    "README.md": "5bfefcdce6eafce3388d1fb200f3b10954cfeac6c7a45fd7dec42687e01ac75d",
    "module.js": "3c07596a6a4796c65ef10ba2bc0805e7f3bc9e4e8fc9970a1307b97e29db1c0a",
    "module.js.LICENSE.txt": "fdbc28c10f3d21976b4bc16464ad7c630538c0c3101347b5fd44af9066f7022b",
    "module.js.map": "c3ac1e8aa98d83c54fd13e43b65e1cf4182a924d2eb23a2f1a6fe40b7785a1bb",
    "plugin.json": "cf26a3afb7c10cd9ae40b5296d04172b5dac927d69a51082e6d085b34341ccc3"
  },
  "time": 1589558058070,
  "keyId": "7e4d0c6a708866e7"
}
-----BEGIN PGP SIGNATURE-----
Version: OpenPGP.js v4.10.1
Comment: https://openpgpjs.org

wqAEARMKAAYFAl6+uyoACgkQfk0ManCIZuc0+QIHdWC0dp7GRRFu3Hgk9tnl
FJnPwM6Y2tTdq7AkpVTTAb3RTFadA8dRmLfajxgHxmDf5yUv9M2M6sa1eTSG
8kJtOlwCB096dXOKsH1IOGQMCY+/xM2081FqbMTvWgN81xrxMoxftQn8z6VC
2nA2Rmt1VStppFVCCUXaq6Y4sFGHQF/yq5oi
=vqUQ
-----END PGP SIGNATURE-----

```

## Verifying a plugin

> When Grafana starts up it discover which plugins to load. For each plugin it verifies the authenticity of it, whether to load it or not based on the state of the plugin signature.

Reviewer: ```suggestion
When Grafana starts it discovers plugins to load. For each discovered plugin it verifies the authenticity of it, and then decides whether to load it or not based on the state of the plugin signature:
```

> 1. If plugin is a core plugin built into Grafana, plugin signature is `internal`.

Reviewer: ... then the plugin signature is `internal`. (and for the following).

Or maybe turn this into a table with two columns, State, Conditions.

Author: Table is maybe better. WIll try it out

Author: Changed to a table

1. If no _MANIFEST.txt_ file is found, plugin signature is `unsigned`.
1. Using the public key (compiled with Grafana), if the digital signature of _MANIFEST.txt_ is not valid, plugin signature is `invalid`.
1. If manifest plugin id or version have been changed, plugin signature is `modified`.
1. If manifest files checksums doesn't match, plugin signature is `modified`.
1. If all earlier steps passes, plugin signature is `valid`.

The plugin signature state can be inspected for each plugin in the plugins listing page (Configuration -> Plugins).

### Backend plugins

> If a [backend plugin]({{< relref "backend/_index.md" >}}) is not signed Grafana will not load/start it. Trying to load a backend plugin with an invalid sigature will write an error message to the Grafana server log `plugin <plugin id> is unsigned`.

Reviewer: ```suggestion
If a [backend plugin]({{< relref "backend/_index.md" >}}) is not signed then Grafana will not load or start it. Trying to load a backend plugin with an invalid signature will result in an error message in Grafana server's log in the format of `plugin <plugin id> is unsigned`.
```

### Allow unsigned plugins

> It's possible to allow unsigned plugins using a configuration setting, but is something we strongly advise against doing. Read more [here]({{< relref "../../installation/configuration.md#allow-loading-unsigned-plugins" >}}) about this setting. Running an unsiged backend plugin will write a warning message to the Grafana server log `Running an unsigned backend plugin   pluginID=<plugin id>`.

Reviewer: Do we also skip verification in dev mode now? If so I think we should mention this as well.

Author: Yeah, good point.

Author: Added a note about this. What do you think?

Reviewer 3: We do not skip verificatoin -- we *do* skip requiring them for backend plugins.  That is, in dev mode if a MANIFEST exists and is invalid it will fail, but it will be allowed to run with no MANIFEST

Author: > That is, in dev mode if a MANIFEST exists and is invalid it will fail, but it will be allowed to run with no MANIFEST

It only fails if it's a backend plugin.

Author: Adding a note about that we don't skip verification

> It's possible to allow unsigned plugins using a configuration setting, but is something we strongly advise against doing. Read more [here]({{< relref "../../installation/configuration.md#allow-loading-unsigned-plugins" >}}) about this setting. Running an unsiged backend plugin will write a warning message to the Grafana server log `Running an unsigned backend plugin   pluginID=<plugin id>`.

Reviewer 2: ```suggestion
While you can allow unsigned plugins using a configuration setting, we strongly advise you not to. If you run an unsigned backend plugin, then Grafana writes a warning message to the server log.

    Running an unsigned backend plugin   pluginID=<plugin id>

For more information on how to allow unsigned backend plugin, refer to [Configuration]({{< relref "../../installation/configuration.md#allow-loading-unsigned-plugins" >}}). 
```
